sql/tests: create sysbench microbenchmark suite

nvb opened the pull request:
This commit adds a new microbenchmark suite which emulates `sysbench`, with subtests for `oltp_read_only`, `oltp_write_only`, and `oltp_read_write`. The structure of the benchmark suite makes it easy to add more subtests in the future (e.g. for `oltp_multi_insert`).

The suite is designed to run the same workload against different levels of the CockroachDB stack, with the initial implementation targeting SQL (and below) through the `sysbenchSQL` driver and KV (and below) through the `sysbenchKV` driver. An example of additional an driver that could be added in the future is `sysbenchPebble`. It would also make sense to test drivers under different configurations. For example, we should add a variant of the `sysbenchKV` driver which disables the local RPC fast-path.

The goal of this suite is to provide developers with a way to quickly measure the performance of CockroachDB against sysbench in a Go microbenchmark environment. This will help with identifying opportunities for performance improvements and with providing a rapid feedback loop for evaluating the effectiveness of performance changes.

The initial benchmark performance looks like:
```
name                        time/op
Sysbench/SQL/OltpReadOnly   2.88ms ±16%
Sysbench/SQL/OltpWriteOnly  2.32ms ±16%
Sysbench/SQL/OltpReadWrite  5.47ms ± 6%
Sysbench/KV/OltpReadOnly     445µs ± 5%
Sysbench/KV/OltpWriteOnly    594µs ± 5%
Sysbench/KV/OltpReadWrite   1.07ms ± 4%

name                        alloc/op
Sysbench/SQL/OltpReadOnly    965kB ± 1%
Sysbench/SQL/OltpWriteOnly   487kB ± 4%
Sysbench/SQL/OltpReadWrite  1.34MB ± 0%
Sysbench/KV/OltpReadOnly     264kB ± 1%
Sysbench/KV/OltpWriteOnly    184kB ± 4%
Sysbench/KV/OltpReadWrite    440kB ± 0%

name                        allocs/op
Sysbench/SQL/OltpReadOnly    6.26k ± 1%
Sysbench/SQL/OltpWriteOnly   3.40k ± 0%
Sysbench/SQL/OltpReadWrite   9.65k ± 0%
Sysbench/KV/OltpReadOnly       650 ± 0%
Sysbench/KV/OltpWriteOnly    1.08k ± 1%
Sysbench/KV/OltpReadWrite    1.73k ± 0%
```

Epic: None
Release Note: None

nvb committed:
sql/tests: create sysbench microbenchmark suite

This commit adds a new microbenchmark suite which emulates `sysbench`, with
subtests for `oltp_read_only`, `oltp_write_only`, and `oltp_read_write`. The
structure of the benchmark suite makes it easy to add more subtests in the
future (e.g. for `oltp_multi_insert`).

The suite is designed to run the same workload against different levels of the
CockroachDB stack, with the initial implementation targeting SQL (and below)
through the `sysbenchSQL` driver and KV (and below) through the `sysbenchKV`
driver. An example of additional an driver that could be added in the future is
`sysbenchPebble`. It would also make sense to test drivers under different
configurations. For example, we should add a variant of the `sysbenchKV` driver
which disables the local RPC fast-path.

The goal of this suite is to provide developers with a way to quickly measure
the performance of CockroachDB against sysbench in a Go microbenchmark
environment. This will help with identifying opportunities for performance
improvements and with providing a rapid feedback loop for evaluating the
effectiveness of performance changes.

The initial benchmark performance looks like:
```
name                        time/op
Sysbench/SQL/oltp_read_only   2.88ms ±16%
Sysbench/SQL/oltp_write_only  2.32ms ±16%
Sysbench/SQL/oltp_read_write  5.47ms ± 6%
Sysbench/KV/oltp_read_only     445µs ± 5%
Sysbench/KV/oltp_write_only    594µs ± 5%
Sysbench/KV/oltp_read_write   1.07ms ± 4%

name                        alloc/op
Sysbench/SQL/oltp_read_only    965kB ± 1%
Sysbench/SQL/oltp_write_only   487kB ± 4%
Sysbench/SQL/oltp_read_write  1.34MB ± 0%
Sysbench/KV/oltp_read_only     264kB ± 1%
Sysbench/KV/oltp_write_only    184kB ± 4%
Sysbench/KV/oltp_read_write    440kB ± 0%

name                        allocs/op
Sysbench/SQL/oltp_read_only    6.26k ± 1%
Sysbench/SQL/oltp_write_only   3.40k ± 0%
Sysbench/SQL/oltp_read_write   9.65k ± 0%
Sysbench/KV/oltp_read_only       650 ± 0%
Sysbench/KV/oltp_write_only    1.08k ± 1%
Sysbench/KV/oltp_read_write    1.73k ± 0%
```

Epic: None
Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133132)
<!-- Reviewable:end -->

nvb committed:
sql/tests: add oltp_begin_commit to sysbench microbenchmark suite

This commit adds a new "begin+commit" variant to the sysbench microbenchmark
suite. This test opens and closes a transaction, without actually performing any
reads or writes in that transaction. It is meant to measure the overhead of
transaction orchestration.

The initial benchmark performance looks like:
```
name                          time/op
Sysbench/SQL/oltp_begin_commit   125µs ± 7%
Sysbench/KV/oltp_begin_commit   1.30µs ± 3%

name                          alloc/op
Sysbench/SQL/oltp_begin_commit  19.2kB ± 0%
Sysbench/KV/oltp_begin_commit   2.50kB ± 0%

name                          allocs/op
Sysbench/SQL/oltp_begin_commit     143 ± 0%
Sysbench/KV/oltp_begin_commit     6.00 ± 0%
```

Epic: None
Release Note: None

tbg commented on pkg/sql/tests/sysbench_test.go:
```diff
@@ -0,0 +1,624 @@
+	"github.com/jackc/pgx/v5"
+)
+
+const (
```
Add a comment here that this corresponds to "official" sysbench.

tbg commented on pkg/sql/tests/sysbench_test.go:
```diff
@@ -0,0 +1,624 @@
+	prep(rng *rand.Rand)
+}
+
+const (
```
Put a permalink to the corresponding stmt in the sysbench source in an inline comment behind each of these queries?

tbg replied:
A comment that points readers at the `sysbenchOltp*` methods for how these are arranged in sysbench workloads could be helpful.

tbg commented on pkg/sql/tests/sysbench_test.go:
```diff
@@ -0,0 +1,624 @@
+	s.Commit()
+}
+
+func BenchmarkSysbench(b *testing.B) {
```
Should we have a flavor that disables the local server optimization? Otherwise we're eliding a bunch of networking/encoding/decoding overhead here.

tbg commented on pkg/sql/tests/sysbench_test.go:
```diff
@@ -0,0 +1,624 @@
+}
+
+func try0(err error) {
+	if err != nil {
```
how confident are we that we don't have to handle errors here? I know, single node cockroach usually doesn't have any, but can we at least fail gracefully? For example, stash the error in an atomic and check that at conclusion time. After all, this stuff will run in CI (and our ambition may be to put it squarely into the critical path once we can justify it).

tbg reviewed (commented):
Very nice, thanks for putting this together so quickly! Flushing out some comments, I still need to do the line-by-line review, though I don't think I'll find much.

For posterity: I collected and attached profiles via

```
./dev bench --ignore-cache --stream-output --bench-mem ./pkg/sql/tests/ --filter BenchmarkSysbench/SQL/OltpReadWrite --test-args '-test.cpuprofile=cpu.pb.gz -test.memprofile=mem.pb.gz -test.benchtime=10000x'
```

[cpu.pb.gz](https://pprof.me/1d9cf99202e7b10246ba916116a07534)
[mem.pb.gz](https://pprof.me/cb2ae7f8054c6bfa2743ba75d679e3f5)

It'll be interesting to compare them with the "real" sysbench on a three-node cluster.

RaduBerinde reviewed (commented):
Very cool!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-:-O9tj7F91g24X0koXaci:b-jdo7t9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @nvanbenschoten, and @tbg)

___
*[`pkg/sql/tests/sysbench_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9th-_4ADQshyjWWNeY:-O9th-_50Eo4CXaUQwhB:b-kcro0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L126)):*
> ```Go
> )
> 
> // sysbenchSQL is SQL-based implementation of sysbenchDriver.
> ```

[nit] WDYT about moving this to a `sysbench_sql_test.go` (and the KV one to `sysbench_kv_test`)?  It will make it easier to navigate the code, especially if we add more drivers.
___
*[`pkg/sql/tests/sysbench_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9thOj97JL-TfEH8_qa:-O9thOj97JL-TfEH8_qb:bfxd2dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L126)):*
> ```Go
> )
> 
> // sysbenchSQL is SQL-based implementation of sysbenchDriver.
> ```

[nit] A bit more to this comment would help someone who is just looking at this code for the first time. "It runs SQL statements against a single node cluster"
___
*[`pkg/sql/tests/sysbench_test.go` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9thf1k68PLdwoVA6oq:-O9thf1k68PLdwoVA6or:b-o761wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L261)):*
> ```Go
> }
> 
> // sysbenchKV is KV-based implementation of sysbenchDriver.
> ```

[nit] "It bypasses the SQL layer and runs the workload directly against the KV layer, on a single node cluster"
___
*[`pkg/sql/tests/sysbench_test.go` line 570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#gh-1812724167:-O9tifOg3y4HLblHBAy6:b-ayntf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L570)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should we have a flavor that disables the local server optimization? Otherwise we're eliding a bunch of networking/encoding/decoding overhead here.
</blockquote></details>

Should we have variants with 3 nodes? (can be a TODO for now)
___
*[`pkg/sql/tests/sysbench_test.go` line 577 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9ti9fC8hCsfwtV-sOa:-O9ti9fC8hCsfwtV-sOb:baxtsx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L577)):*
> ```Go
> 		// TODO(nvanbenschoten): add a pebble-level implementation.
> 	} {
> 		sysTyp := runtime.FuncForPC(reflect.ValueOf(sysFn).Pointer()).Name()
> ```

[nit] This is unnecessarily complicated to read and doesn't allow flexibility to make the names more friendly. We can just do

```
drivers := []struct{
  name string
  constructorFn func..
 }{
  { name: "SQL", constructorFn: newSysbenchSQL },
  { name: "KV", constructorFn: newSysbenchKV },
}
```
and similar for the ops.


<!-- Sent from Reviewable.io -->

tbg commented on pkg/sql/tests/sysbench_test.go:
```diff
@@ -0,0 +1,630 @@
+	// NOTE: disabling background work makes the benchmark more predictable, but
+	// also moderately less realistic.
+	disableBackgroundWork(st)
+	s := serverutils.StartServerOnly(b, base.TestServerArgs{Settings: st})
```
https://github.com/cockroachdb/cockroach/issues/133307#issuecomment-2438199014 suggests this might not be as close to the real thing as we like. Leaving this comment as a discussion placeholder.

nvb committed:
roachpb: panic with error in Value.InitChecksum

This gets us a stack trace in the panic value.

Epic: None
Release note: None

nvb committed:
sql/tests: add oltp_point_select to sysbench microbenchmark suite

This commit adds a new "point_select" variant to the sysbench microbenchmark
suite. Like the real sysbench version of this test, this new test performs a
single point select, outside of an explicit transaction.

The initial benchmark performance looks like:
```
name                            time/op
Sysbench/SQL/oltp_point_select   159µs ± 4%
Sysbench/KV/oltp_point_select   18.5µs ± 2%

name                            alloc/op
Sysbench/SQL/oltp_point_select  34.1kB ± 0%
Sysbench/KV/oltp_point_select   5.15kB ± 1%

name                            allocs/op
Sysbench/SQL/oltp_point_select     309 ± 0%
Sysbench/KV/oltp_point_select     37.0 ± 0%
```

Epic: None
Release Note: None

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133132)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @tbg)

___
*[`pkg/sql/tests/sysbench_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9thOj97JL-TfEH8_qa:-OAo5Azs4ScAmjenYeNU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L126)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] A bit more to this comment would help someone who is just looking at this code for the first time. "It runs SQL statements against a single node cluster"
</blockquote></details>

Done.
___
*[`pkg/sql/tests/sysbench_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9th-_4ADQshyjWWNeY:-OAo5Cv483rlQKXADgZa:b-qkxgb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L126)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] WDYT about moving this to a `sysbench_sql_test.go` (and the KV one to `sysbench_kv_test`)?  It will make it easier to navigate the code, especially if we add more drivers.
</blockquote></details>

So far I've found it nice to have this all in the same file. If that changes, we can extract out some driver implementation files.
___
*[`pkg/sql/tests/sysbench_test.go` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9thf1k68PLdwoVA6oq:-OAo5CODET5LMYECtp7m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L261)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "It bypasses the SQL layer and runs the workload directly against the KV layer, on a single node cluster"
</blockquote></details>

Done.
___
*[`pkg/sql/tests/sysbench_test.go` line 577 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-O9ti9fC8hCsfwtV-sOa:-OAo5OzU9GruFanYD19o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f113414690f68a542cf66e032a0c20737a6e560/pkg/sql/tests/sysbench_test.go#L577)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This is unnecessarily complicated to read and doesn't allow flexibility to make the names more friendly. We can just do

```
drivers := []struct{
  name string
  constructorFn func..
 }{
  { name: "SQL", constructorFn: newSysbenchSQL },
  { name: "KV", constructorFn: newSysbenchKV },
}
```
and similar for the ops.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Nice!  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 4 files at r1, 3 of 3 files at r3, 3 of 3 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-:-OAscTas5M7By93LgEvd:b-ypbbbp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @tbg)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTRs!

bors r+

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133132)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[`pkg/sql/tests/sysbench_test.go` line 669 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-OAsoSY86agrbEFm6Vhk:-OAsoSY9Ba_QcL1gaL03:b-t1al65) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ffe49895ce529375f6fcf08c62ac4fa724842b/pkg/sql/tests/sysbench_test.go#L669)):*
> ```Go
> 
> func disableProfiling() {
> 	runtime.MemProfileRate = 0
> ```

The `MemProfileRate` doc says this:
```
The tools that process the memory profiles assume that the profile rate is constant across the lifetime of the program and equal to the current value. Programs that change the memory profiling rate should do so just once, as early as possible in the execution of the program (for example, at the beginning of main).

```
___
*[`pkg/sql/tests/sysbench_test.go` line 676 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-OAsofCy5iECSAdkZIXV:-OAsofCy5iECSAdkZIXW:bp0yyxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ffe49895ce529375f6fcf08c62ac4fa724842b/pkg/sql/tests/sysbench_test.go#L676)):*
> ```Go
> func enableProfiling() {
> 	runtime.GC()
> 	runtime.MemProfileRate = 1
> ```

MemProfileRate is 512KB by default.. A value of 1 means we sample every allocation. Would the CPU profile still be useful when running with this kind of overhead?
___
*[`pkg/sql/tests/sysbench_test.go` line 677 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/133132#-OAsp4yk1N3I3imnZLAv:-OAsp4yk1N3I3imnZLAw:b-27j41i) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ffe49895ce529375f6fcf08c62ac4fa724842b/pkg/sql/tests/sysbench_test.go#L677)):*
> ```Go
> 	runtime.GC()
> 	runtime.MemProfileRate = 1
> 	runtime.SetMutexProfileFraction(1)
> ```

Where are we getting these values? Would it be better to just record the old values from `disableProfiling` and restore those?

Maybe an easier way would be to do the actual workload inside a child `b.Run` benchmark.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r-

nvanbenschoten commented:
@RaduBerinde had some good points about the attempt to isolate memory, mutex, and block profiling to the workload portion of the benchmark. Since that needs some work, I've shaved off that commit for now to get the rest of this PR merged. I'll put up a separate PR once that part is ready.

bors r+
